File: chap5/sec5.6/PIDC.m
function [t,u]=PIDC(lambda,tspan,u0,N,M,K)
%PIDC Solves the Dahlquist equation with Parallel IDC.
%   [t,u]=PIDC(lambda,tspan,u0,N,M,K) solves the Dahlquist equation
%   with lambda on tspan=[tBeg, tEnd], starting from the initial value
%   u0. It uses N parallel IDC time steps, with M points per time
%   step (including the left and right boundary point). Then it computes
%   the prediction (K=0) using Backward Euler, and corrects it using K
%   Backward Euler sweeps.

nPoints=N*(M-1)+1;
t=linspace(tspan(1),tspan(2),nPoints);      % Time grid for all intervals
dt=t(2)-t(1); uStep=ones(1,K+1)*u0;         % Initial guess
for n=1:N                                   
  u{n}{1}(1)=uStep(1);
  for m=1:M-1
    u{n}{1}(m+1)=1/(1-dt*lambda)*u{n}{1}(m);
  end
  for k=1:K                                 % Sweep iterations
    u{n}{k+1}(1)=uStep(k+1);
    integrand=lambda*u{n}{k};
    for m=1:M-1
      weights=LagrangeWeights(M);           % Integral term
      integral=(M-1)*dt*sum(weights(m, :).*integrand);
      u{n}{k+1}(m+1)=1/(1-dt*lambda)*( ...  % Sweep correction
        u{n}{k+1}(m)-dt*lambda*u{n}{k}(m+1)+integral);
    end
  end
  for k=1:K+1
    uStep(k)=u{n}{k}(M);                    % Initial values for next step
  end
end
uAll=u0;                                    % Combine each step
for n=1:N
  uAll=[uAll,u{n}{K+1}(2:end)];
end
u=uAll;

File: chap5/sec5.6/PIDC.py
import numpy as np
from LagrangeWeights import lagrangeWeights

def PIDC(lam, tspan, u0, N, M, K):
    """
    Solves the Dahlquist equation with Parallel IDC.

    This function solves the Dahlquist equation with lambda on tspan=[tBeg, tEnd],
    starting from the initial value u0. It uses N parallel IDC time steps, with M points
    per time step (including the left and right boundary point). Then it computes the
    prediction (K=0) using Backward Euler, and corrects it using K Backward Euler sweeps.

    Parameters
    ----------
    lam : complex
        Parameter lambda in the Dahlquist equation.
    tspan : tuple
        Time interval (tBeg, tEnd).
    u0 : float
        Initial value.
    N : int
        Number of parallel IDC time steps.
    M : int
        Number of points per time step.
    K : int
        Number of correction sweeps.

    Returns
    -------
    t : array
        Time grid.
    u : array
        Solution values at time points.
    """
    nPoints = N * (M - 1) + 1
    t = np.linspace(tspan[0], tspan[1], nPoints)
    dt = t[1] - t[0]
    uStep = np.ones(K + 1, dtype=complex) * u0

    u = {}
    for n in range(N):
        u[n] = {}
        u[n][1] = np.zeros(M, dtype=complex)
        u[n][1][0] = uStep[0]

        # Initial guess using Backward Euler
        for m in range(M - 1):
            u[n][1][m + 1] = 1 / (1 - dt * lam) * u[n][1][m]

        # Sweep iterations
        for k in range(K):
            u[n][k + 2] = np.zeros(M, dtype=complex)
            u[n][k + 2][0] = uStep[k + 1]
            integrand = lam * u[n][k + 1]

            for m in range(M - 1):
                weights = lagrangeWeights(M)
                integral = (M - 1) * dt * sum(weights[m] * integrand)
                u[n][k + 2][m + 1] = 1 / (1 - dt * lam) * (
                    u[n][k + 2][m] - dt * lam * u[n][k + 1][m + 1] + integral
                )

        # Update initial values for the next step
        for k in range(K + 1):
            uStep[k] = u[n][k + 1][M - 1]

    # Combine each step
    uAll = [u0]
    for n in range(N):
        uAll.extend(u[n][K + 1][1:])

    return t, np.array(uAll)
